salt/modules/virt.py

```diff
@@ -735,10 +736,58 @@ def _gen_xml(
     context = {
         "hypervisor": hypervisor,
         "name": name,
-        "cpu": six.text_type(cpu),
         "mem": six.text_type(mem),
     }
 
+    if isinstance(cpu, int):
+        context["cpu"].update(
+            {"maximum": six.text_type(cpu)}, {"current": six.text_type(cpu)}
+        )
+    elif isinstance(cpu, dict):
+        context["cpu"] = cpu if cpu else {}
+
+    # preprocess the cpu context, make attributes directly usable in jinja template.
+    if context["cpu"]:
+        for key in context["cpu"]:
+            # make vcpu attributes directly accessible
+            if key in ["placement", "cpuset", "current"]:
+                context["cpu"].update({key: "{}='{}'".format(key, context["cpu"][key])})
+            elif key == "vcpus":
+                for vcpu_id in context["cpu"]["vcpus"]:
+                    for vcpu_attr in context["cpu"]["vcpus"][vcpu_id]:
+                        context["cpu"]["vcpus"][vcpu_id].update(
+                            {
+                                vcpu_attr: "{}='{}'".format(
+                                    vcpu_attr,
+                                    context["cpu"]["vcpus"][vcpu_id][vcpu_attr],
+                                )
+                            }
+                        )
+            elif key in ["attrib", "model", "topology", "cache", "feature"]:
```

Comment:
It's possible to get rid of `attrib`  But it might make the jinja code look even more ugly and less readable. This is how I handle `cpu attribute` in jinja templates  `<cpu {% for attr in cpu.attrib %} {{ cpu.attrib[attr] + ''}} {%- endfor %}>` There are 3 possible cpu attributes and 2 of them are optional, I can't think of a neater way of handle it. Do you have better suggestions?

Reply:
Actually, i just had an idea of get rid of attrib without changing the jinja template :)